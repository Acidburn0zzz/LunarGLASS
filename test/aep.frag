#version 310 es

precision highp float;

#extension GL_ANDROID_extension_pack_es31a : enable

// - GL_KHR_blend_equation_advanced
// + GL_OES_sample_variables
// - GL_OES_shader_multisample_interpolation
// + GL_EXT_geometry_shader
// + GL_EXT_shader_io_blocks

in inName {
    vec4 color;
} inInst;

out vec4 color;

int foo_GS();
void goodSample();
vec3 interp();

void main()
{
    color = inInst.color;
    color *= float(foo_GS());
    goodSample();
    color.xyz += interp();
}

// GL_OES_geometry_shader

highp int foo_GS()
{
    highp int l = gl_Layer;
    highp int p = gl_PrimitiveID;
    return l + p;
}


// GL_OES_sample_variables

void goodSample()
{
    lowp     int  a1 = gl_SampleID;
    mediump  vec2 a2 = gl_SamplePosition; 
    highp    int  a3 = gl_SampleMaskIn[0];
    gl_SampleMask[0] = a3;
    mediump int n1 = gl_MaxSamples;
    mediump int n2 = gl_NumSamples;
    gl_SampleMask[0] += n1 + n2 + a1;
    color.xy += a2;
}

// GL_OES_shader_multisample_interpolation

#extension GL_OES_shader_multisample_interpolation : enable

in float scalarIn;

sample in vec4 colorSampIn;
flat sample in vec4 colorfsi;
sample in vec3 sampInArray;  // ??: add the array: "[4]", currently makes two "sample" come out
uniform int i;

vec3 interp()
{
    float res;
    vec3 res3;

    res   = interpolateAtCentroid(scalarIn);
    res3  = interpolateAtCentroid(sampInArray/*[2]*/);

    res3 += interpolateAtSample(sampInArray/*[i]*/, 0);
    res  += interpolateAtSample(scalarIn, 1);

    res3 += interpolateAtOffset(sampInArray/*[2]*/, vec2(0.2));
    res  += interpolateAtOffset(scalarIn, vec2(0.2));

    return res * res3;
}

layout(blend_support_darken) out;
layout(blend_support_all_equations) out;
