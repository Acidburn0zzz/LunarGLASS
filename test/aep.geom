#version 310 es

// + GL_EXT_geometry_shader
// + GL_EXT_geometry_point_size

#extension GL_ANDROID_extension_pack_es31a : enable

precision mediump float;

in fromVertex {
    in vec3 color;
} fromV[];

in vec4 nonBlockUnsized[];

out toFragment {
    out vec3 color;
} toF;

out fromVertex {
    centroid vec3 color;
    int len;
    int inv;
};

layout(triangle_strip) out;
layout(max_vertices = 200) out;
layout(lines_adjacency) in;
layout(invocations = 4) in;

centroid in vec3 centr[];

void pointSize2();

void main()
{
    EmitVertex();
    EndPrimitive();

    color = fromV[0].color + centr[1];
    gl_Position = gl_in[0].gl_Position;

    gl_PrimitiveID = gl_PrimitiveIDIn;
    gl_Layer = 2;

    len = fromV.length();     // 4: lines_adjacency
    inv = gl_InvocationID;

    pointSize2();
}


#extension GL_OES_geometry_point_size : enable

void pointSize2()
{
    highp float ps = gl_in[3].gl_PointSize;
    gl_PointSize = ps;
}
