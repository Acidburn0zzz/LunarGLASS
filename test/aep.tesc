#version 310 es

#extension GL_ANDROID_extension_pack_es31a : enable

// GL_EXT_tessellation_shader: patch
// GL_EXT_tessellation_point_size
// GL_OES_primitive_bounding_box

layout(vertices = 4) out;
int outa[gl_out.length()];

patch out vec4 patchOut;

void bb();
void pointSize();
float foop();

void main()
{
    barrier();

    vec4 p = gl_in[1].gl_Position;

    int pvi = gl_PatchVerticesIn;
    int pid = gl_PrimitiveID;
    int iid = gl_InvocationID;

    gl_out[1].gl_Position = p;

    gl_TessLevelOuter[3] = 3.2;
    gl_TessLevelInner[1] = 1.3;
    bb();
    pointSize();
    patchOut = vec4(0.5);
    patchOut *= foop();
}

in vec2 inb[];
in vec2 ind[gl_MaxPatchVertices];

#extension GL_ARB_separate_shader_objects : enable

layout(location = 3) in vec4 ivla[];
layout(location = 4) in vec4 ivlb[];

layout(location = 3) out vec4 ovla[];
layout(location = 4) out vec4 ovlb[];

patch out pinbn {
    int a;
} pinbi;

centroid out vec3 myColor2[];
centroid in vec3 centr[];
out float okaySize[4];

#extension GL_OES_tessellation_point_size : enable

// GL_EXT_tessellation_point_size

void pointSize()
{
    gl_out[1].gl_PointSize = gl_in[1].gl_PointSize;
}

// for testing with gpu_shader5
in vec3 inv[];
precise vec3 pv3;

float foop()
{
    precise float d;

    pv3 = inv[3];
    pv3 *= pv3;
    pv3 = fma(pv3, pv3, pv3);
    d = fma(pv3.x, pv3.y, pv3.z);

    return d;
}

// GL_OES_primitive_bounding_box

void bb()
{
    gl_BoundingBoxOES[0] = vec4(0.0);
    gl_BoundingBoxOES[1] = vec4(1.0);
}
