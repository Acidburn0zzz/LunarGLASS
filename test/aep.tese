#version 310 es

#extension GL_ANDROID_extension_pack_es31a : enable

// - GL_EXT_tessellation_shader
// + GL_EXT_tessellation_point_size

layout(quads, ccw) in;  // - no ccw

layout(fractional_odd_spacing) in;    

layout(point_mode) in;

patch in vec4 patchIn;

void pointSize();

void main()
{
    vec4 p = gl_in[1].gl_Position;

    int pvi = gl_PatchVerticesIn;
    int pid = gl_PrimitiveID;
    vec3 tc = gl_TessCoord;
    float tlo = gl_TessLevelOuter[3];
    float tli = gl_TessLevelInner[1];

    gl_Position = p;
    pointSize();
}

in vec2 inb[];
in vec2 ind[gl_MaxPatchVertices];

in testblb {
    int f;
} blb[];

in testbld {
    int f;
} bld[gl_MaxPatchVertices];

layout(location = 23) in vec4 ivla[];
layout(location = 24) in vec4 ivlb[];

layout(location = 23) out vec4 ovla[2];

patch in pinbn {
    int a;
} pinbi;

centroid out vec3 myColor2;
centroid in vec3 centr[];

#extension GL_OES_tessellation_point_size : enable

void pointSize()
{
    gl_PointSize = gl_in[1].gl_PointSize;
}
