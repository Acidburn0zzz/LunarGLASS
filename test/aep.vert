#version 310 es

#extension GL_ANDROID_extension_pack_es31a : enable

// GL_OES_shader_image_atomic
// GL_OES_texture_storage_multisample_2d_array
// GL_EXT_gpu_shader5
// GL_EXT_shader_io_blocks
// GL_EXT_texture_buffer
// GL_EXT_texture_cube_map_array

uniform sampler2D sArray[4];
uniform ubName { vec2 p; } ubInst[4];
uniform writeonly lowp image2D iArray[5];
uniform int index;

const ivec2 constOffsets[4] = ivec2[4](ivec2(1), ivec2(2), ivec2(3), ivec2(4));

in vec2 inf, ing, inch;

out outName {
    vec4 color;
} outInst;

vec4 bufferT();
vec4 CAT();
vec4 MSA();
void goodImageAtom();

void main()
{
    // GL_EXT_gpu_shader5
    precise vec2 p = fma(inf, ing, inch);
    iArray[index];
    vec4 color = vec4(0.0);
    color += textureGatherOffset(sArray[index], ubInst[index].p, ivec2(inf));
    color += textureGatherOffsets(sArray[index], p, constOffsets);
    outInst.color = color;
    outInst.color += bufferT();
    outInst.color += CAT();
    outInst.color += MSA();
    goodImageAtom();
}

// GL_EXT_texture_buffer

uniform highp samplerBuffer  bufSamp1;          
uniform highp isamplerBuffer bufSamp2;          
uniform highp usamplerBuffer bufSamp3;          
uniform highp writeonly imageBuffer    bufSamp4;
uniform highp writeonly iimageBuffer   bufSamp5;
uniform highp writeonly uimageBuffer   bufSamp6;

vec4 bufferT()
{
    vec4 v = vec4(1.0);

    highp int s1;
    s1 = textureSize(bufSamp1);
    v *= vec4(s1);
    s1 = textureSize(bufSamp2);
    v *= vec4(s1);
    s1 = textureSize(bufSamp3);
    v *= vec4(s1);

    s1 = imageSize(bufSamp4);
    v *= vec4(s1);
    s1 = imageSize(bufSamp5);
    v *= vec4(s1);
    s1 = imageSize(bufSamp6);
    v *= vec4(s1);
    
    v *= texelFetch(bufSamp1, s1);
    v *= vec4(texelFetch(bufSamp2, s1));
    v *= vec4(texelFetch(bufSamp3, s1));

    return v;
}

// GL_OES_texture_cube_map_array

uniform highp writeonly imageCubeArray  CA1;
uniform highp writeonly iimageCubeArray CA2;
uniform highp writeonly uimageCubeArray CA3;
uniform highp samplerCubeArray          CA4;
uniform highp samplerCubeArrayShadow    CA5;
uniform highp isamplerCubeArray         CA6;
uniform highp usamplerCubeArray         CA7;

vec4 CAT()
{
    highp ivec3 iv = ivec3(0);
    iv += textureSize(CA4, 1);
    iv += textureSize(CA5, 1);
    iv += textureSize(CA6, 1);
    iv += textureSize(CA7, 1);

    iv += imageSize(CA1);
    iv += imageSize(CA2);
    iv += imageSize(CA3);
    
    vec4 v = vec4(iv, 1.0);
    v *= texture(CA4, vec4(0.5));
    v *= texture(CA5, vec4(0.5), 3.0);
    v *= vec4(texture(CA6, vec4(0.5)));
    v *= vec4(texture(CA7, vec4(0.5)));

    v *= textureLod(CA4, vec4(0.5), 0.24);
    v *= vec4(textureLod(CA6, vec4(0.5), 0.26));
    v *= vec4(textureLod(CA7, vec4(0.5), 0.27));

    v *= textureGrad(CA4, vec4(0.5), vec3(0.1), vec3(0.2));
    v *= vec4(textureGrad(CA6, vec4(0.5), vec3(0.1), vec3(0.2)));
    v *= vec4(textureGrad(CA7, vec4(0.5), vec3(0.1), vec3(0.2)));

    v *= textureGather(CA4, vec4(0.5));
    v *= textureGather(CA4, vec4(0.5), 2);
    v *= vec4(textureGather(CA6, vec4(0.5)));
    v *= vec4(textureGather(CA6, vec4(0.5), 1));
    v *= vec4(textureGather(CA7, vec4(0.5)));
    v *= vec4(textureGather(CA7, vec4(0.5), 0));

    v *= textureGather(CA5, vec4(0.5), 2.5);

    return v;
}

// GL_OES_texture_storage_multisample_2d_array

uniform highp sampler2DMSArray  samp2DMSA;
uniform highp isampler2DMSArray samp2DMSAi;
uniform highp usampler2DMSArray samp2DMSAu;

vec4 MSA()
{
    highp ivec3 iv = ivec3(0);
    iv += textureSize(samp2DMSA);
    iv += textureSize(samp2DMSAi);
    iv += textureSize(samp2DMSAu);
    
    vec4 v = vec4(iv, 1.0);
    v *= texelFetch(samp2DMSA, ivec3(5), 2);
    v *= vec4(texelFetch(samp2DMSAi, ivec3(5), 2));
    v *= vec4(texelFetch(samp2DMSAu, ivec3(5), 2));

    return v;
}

// GL_OES_shader_image_atomic

uniform layout(r32f)  highp  image2D im2Df;
uniform layout(r32ui) highp uimage2D im2Du;
uniform layout(r32i)  highp iimage2D im2Di;
uniform ivec2 P;

void goodImageAtom()
{
    float datf = 1.8;
    int dati = 4;
    uint datu = 7u;

    imageAtomicAdd(     im2Di, P, P.x);
    imageAtomicAdd(     im2Du, P, datu);
    imageAtomicMin(     im2Di, P, dati);
    imageAtomicMin(     im2Du, P, datu);
    imageAtomicMax(     im2Di, P, dati);
    imageAtomicMax(     im2Du, P, datu);
    imageAtomicAnd(     im2Di, P, dati);
    imageAtomicAnd(     im2Du, P, datu);
    imageAtomicOr(      im2Di, P, dati);
    imageAtomicOr(      im2Du, P, datu);
    imageAtomicXor(     im2Di, P, dati);
    imageAtomicXor(     im2Du, P, datu);
    imageAtomicExchange(im2Di, P, dati);
    imageAtomicExchange(im2Du, P, datu);
    imageAtomicExchange(im2Df, P, datf);
    imageAtomicCompSwap(im2Di, P,  3, dati);
    imageAtomicCompSwap(im2Du, P, 5u, datu);
}
