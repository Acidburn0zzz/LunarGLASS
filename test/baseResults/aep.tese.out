
Top IR:
; ModuleID = 'Glslang'

%gl_PerVertex = type { <4 x float>, float }
%gl_PerVertex.0 = type { <4 x float>, float }
%testblb = type { i32 }
%testbld = type { i32 }
%pinbn = type { i32 }

@gl_in = global [32 x %gl_PerVertex] zeroinitializer
@gl_PatchVerticesIn = global i32 0
@gl_PrimitiveID = global i32 0
@gl_TessCoord = global <3 x float> zeroinitializer
@gl_TessLevelOuter = global [4 x float] zeroinitializer
@gl_TessLevelInner = global [2 x float] zeroinitializer
@"anon@1" = global %gl_PerVertex.0 zeroinitializer
@patchIn = global <4 x float> zeroinitializer
@inb = global [32 x <2 x float>] zeroinitializer
@ind = global [32 x <2 x float>] zeroinitializer
@blb = global [32 x %testblb] zeroinitializer
@bld = global [32 x %testbld] zeroinitializer
@ivla = global [32 x <4 x float>] zeroinitializer
@ivlb = global [32 x <4 x float>] zeroinitializer
@ovla = global [2 x <4 x float>] zeroinitializer
@pinbi = global %pinbn zeroinitializer
@myColor2 = global <3 x float> zeroinitializer
@centr = global [32 x <3 x float>] zeroinitializer

define fastcc void @main() {
entry:
  %tli = alloca float
  %tlo = alloca float
  %tc = alloca <3 x float>
  %pid = alloca i32
  %pvi = alloca i32
  %p = alloca <4 x float>
  br label %mainBody

mainBody:                                         ; preds = %entry
  %p1 = load <4 x float>* getelementptr inbounds ([32 x %gl_PerVertex]* @gl_in, i32 0, i32 1, i32 0)
  store <4 x float> %p1, <4 x float>* %p
  %pvi2 = load i32* @gl_PatchVerticesIn
  store i32 %pvi2, i32* %pvi
  %pid3 = load i32* @gl_PrimitiveID
  store i32 %pid3, i32* %pid
  %tc4 = load <3 x float>* @gl_TessCoord
  store <3 x float> %tc4, <3 x float>* %tc
  %tlo5 = load float* getelementptr inbounds ([4 x float]* @gl_TessLevelOuter, i32 0, i32 3)
  store float %tlo5, float* %tlo
  %tli6 = load float* getelementptr inbounds ([2 x float]* @gl_TessLevelInner, i32 0, i32 1)
  store float %tli6, float* %tli
  %0 = load <4 x float>* %p
  store <4 x float> %0, <4 x float>* getelementptr inbounds (%gl_PerVertex.0* @"anon@1", i32 0, i32 0)
  call void @"pointSize("()
  br label %stage-epilogue

stage-epilogue:                                   ; preds = %mainBody
  br label %stage-exit

stage-exit:                                       ; preds = %stage-epilogue
  ret void
}

; Function Attrs: alwaysinline
define internal fastcc void @"pointSize("() #0 {
entry:
  %0 = load float* getelementptr inbounds ([32 x %gl_PerVertex]* @gl_in, i32 0, i32 1, i32 1)
  store float %0, float* getelementptr inbounds (%gl_PerVertex.0* @"anon@1", i32 0, i32 1)
  ret void
}

attributes #0 = { alwaysinline }

!gla.inputPrimitive = !{!0}
!gla.vertexSpacing = !{!1}
!gla.vertexOrder = !{!2}
!gla.pointMode = !{!3}
!gla.entrypoint = !{!4}
!gla.inputs = !{!5, !13, !15, !17, !19, !21, !23, !25, !27, !29, !34, !37, !39, !41, !44}
!gla.outputs = !{!46, !48, !49}
!gla.noStaticUse = !{!23, !25, !27, !29, !34, !37, !39, !48, !41, !49, !44}

!0 = metadata !{i32 8}
!1 = metadata !{i32 3}
!2 = metadata !{i32 2}
!3 = metadata !{i32 1}
!4 = metadata !{metadata !"main", i32 15}
!5 = metadata !{metadata !"gl_in", i32 17, [32 x %gl_PerVertex]* @gl_in_typeProxy, metadata !6, metadata !7}
!6 = metadata !{i32 0, i32 0, i32 1024, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!7 = metadata !{metadata !"gl_PerVertex", metadata !8, metadata !"gl_Position", metadata !9, metadata !"gl_PointSize", metadata !11}
!8 = metadata !{i32 0, i32 0, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!9 = metadata !{metadata !"", metadata !10}
!10 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 9, i32 -1, i32 0, i32 -1}
!11 = metadata !{metadata !"", metadata !12}
!12 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 10, i32 -1, i32 0, i32 -1}
!13 = metadata !{metadata !"gl_PatchVerticesIn", i32 1, i32* @gl_PatchVerticesIn_typeProxy, metadata !14}
!14 = metadata !{i32 0, i32 3, i32 1088, null, i32 0, i32 34, i32 -1, i32 0, i32 -1}
!15 = metadata !{metadata !"gl_PrimitiveID", i32 1, i32* @gl_PrimitiveID_typeProxy, metadata !16}
!16 = metadata !{i32 0, i32 3, i32 1089, null, i32 0, i32 31, i32 -1, i32 0, i32 -1}
!17 = metadata !{metadata !"gl_TessCoord", i32 1, <3 x float>* @gl_TessCoord_typeProxy, metadata !18}
!18 = metadata !{i32 0, i32 3, i32 1090, null, i32 0, i32 37, i32 -1, i32 0, i32 -1}
!19 = metadata !{metadata !"gl_TessLevelOuter", i32 1, [4 x float]* @gl_TessLevelOuter_typeProxy, metadata !20}
!20 = metadata !{i32 0, i32 3, i32 1091, null, i32 3, i32 35, i32 -1, i32 0, i32 -1}
!21 = metadata !{metadata !"gl_TessLevelInner", i32 1, [2 x float]* @gl_TessLevelInner_typeProxy, metadata !22}
!22 = metadata !{i32 0, i32 3, i32 1095, null, i32 3, i32 36, i32 -1, i32 0, i32 -1}
!23 = metadata !{metadata !"patchIn", i32 1, <4 x float>* @patchIn_typeProxy, metadata !24}
!24 = metadata !{i32 0, i32 3, i32 1099, null, i32 3, i32 0, i32 -1, i32 0, i32 -1}
!25 = metadata !{metadata !"inb", i32 1, [32 x <2 x float>]* @inb_typeProxy, metadata !26}
!26 = metadata !{i32 0, i32 3, i32 1100, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!27 = metadata !{metadata !"ind", i32 1, [32 x <2 x float>]* @ind_typeProxy, metadata !28}
!28 = metadata !{i32 0, i32 3, i32 1132, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!29 = metadata !{metadata !"blb", i32 17, [32 x %testblb]* @blb_typeProxy, metadata !30, metadata !31}
!30 = metadata !{i32 0, i32 0, i32 1164, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!31 = metadata !{metadata !"testblb", metadata !8, metadata !"f", metadata !32}
!32 = metadata !{metadata !"", metadata !33}
!33 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!34 = metadata !{metadata !"bld", i32 17, [32 x %testbld]* @bld_typeProxy, metadata !35, metadata !36}
!35 = metadata !{i32 0, i32 0, i32 1196, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!36 = metadata !{metadata !"testbld", metadata !8, metadata !"f", metadata !32}
!37 = metadata !{metadata !"ivla", i32 1, [32 x <4 x float>]* @ivla_typeProxy, metadata !38}
!38 = metadata !{i32 0, i32 3, i32 23, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!39 = metadata !{metadata !"ivlb", i32 1, [32 x <4 x float>]* @ivlb_typeProxy, metadata !40}
!40 = metadata !{i32 0, i32 3, i32 24, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!41 = metadata !{metadata !"pinbi", i32 17, %pinbn* @pinbi_typeProxy, metadata !42, metadata !43}
!42 = metadata !{i32 0, i32 0, i32 1228, null, i32 3, i32 0, i32 -1, i32 0, i32 -1}
!43 = metadata !{metadata !"pinbn", metadata !8, metadata !"a", metadata !32}
!44 = metadata !{metadata !"centr", i32 1, [32 x <3 x float>]* @centr_typeProxy, metadata !45}
!45 = metadata !{i32 0, i32 3, i32 1230, null, i32 512, i32 0, i32 -1, i32 0, i32 -1}
!46 = metadata !{metadata !"", i32 16, %gl_PerVertex.0* @"anon@1_typeProxy", metadata !47, metadata !7}
!47 = metadata !{i32 0, i32 0, i32 1097, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!48 = metadata !{metadata !"ovla", i32 7, [2 x <4 x float>]* @ovla_typeProxy, metadata !38}
!49 = metadata !{metadata !"myColor2", i32 7, <3 x float>* @myColor2_typeProxy, metadata !50}
!50 = metadata !{i32 0, i32 3, i32 1229, null, i32 512, i32 0, i32 -1, i32 0, i32 -1}


Bottom IR:
; ModuleID = 'Glslang'

%gl_PerVertex = type { <4 x float>, float }
%gl_PerVertex.0 = type { <4 x float>, float }
%testblb = type { i32 }
%testbld = type { i32 }
%pinbn = type { i32 }

@gl_in = global [32 x %gl_PerVertex] zeroinitializer
@gl_PatchVerticesIn = global i32 0
@gl_PrimitiveID = global i32 0
@gl_TessCoord = global <3 x float> zeroinitializer
@gl_TessLevelOuter = global [4 x float] zeroinitializer
@gl_TessLevelInner = global [2 x float] zeroinitializer
@"anon@1" = global %gl_PerVertex.0 zeroinitializer
@patchIn = global <4 x float> zeroinitializer
@inb = global [32 x <2 x float>] zeroinitializer
@ind = global [32 x <2 x float>] zeroinitializer
@blb = global [32 x %testblb] zeroinitializer
@bld = global [32 x %testbld] zeroinitializer
@ivla = global [32 x <4 x float>] zeroinitializer
@ivlb = global [32 x <4 x float>] zeroinitializer
@ovla = global [2 x <4 x float>] zeroinitializer
@pinbi = global %pinbn zeroinitializer
@myColor2 = global <3 x float> zeroinitializer
@centr = global [32 x <3 x float>] zeroinitializer

define fastcc void @main() {
entry:
  %gla_constGEP = getelementptr [32 x %gl_PerVertex]* @gl_in, i32 0, i32 1, i32 0
  %p1 = load <4 x float>* %gla_constGEP, align 16
  %gla_constGEP7 = getelementptr %gl_PerVertex.0* @"anon@1", i32 0, i32 0
  store <4 x float> %p1, <4 x float>* %gla_constGEP7, align 16
  %gla_constGEP8 = getelementptr [32 x %gl_PerVertex]* @gl_in, i32 0, i32 1, i32 1
  %0 = load float* %gla_constGEP8, align 16
  %gla_constGEP9 = getelementptr %gl_PerVertex.0* @"anon@1", i32 0, i32 1
  store float %0, float* %gla_constGEP9, align 16
  br label %stage-epilogue

stage-epilogue:                                   ; preds = %entry
  br label %stage-exit

stage-exit:                                       ; preds = %stage-epilogue
  ret void
}

!gla.inputPrimitive = !{!0}
!gla.vertexSpacing = !{!1}
!gla.vertexOrder = !{!2}
!gla.pointMode = !{!3}
!gla.entrypoint = !{!4}
!gla.inputs = !{!5, !13, !15, !17, !19, !21, !23, !25, !27, !29, !34, !37, !39, !41, !44}
!gla.outputs = !{!46, !48, !49}
!gla.noStaticUse = !{!23, !25, !27, !29, !34, !37, !39, !48, !41, !49, !44}

!0 = metadata !{i32 8}
!1 = metadata !{i32 3}
!2 = metadata !{i32 2}
!3 = metadata !{i32 1}
!4 = metadata !{metadata !"main", i32 15}
!5 = metadata !{metadata !"gl_in", i32 17, [32 x %gl_PerVertex]* @gl_in_typeProxy, metadata !6, metadata !7}
!6 = metadata !{i32 0, i32 0, i32 1024, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!7 = metadata !{metadata !"gl_PerVertex", metadata !8, metadata !"gl_Position", metadata !9, metadata !"gl_PointSize", metadata !11}
!8 = metadata !{i32 0, i32 0, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!9 = metadata !{metadata !"", metadata !10}
!10 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 9, i32 -1, i32 0, i32 -1}
!11 = metadata !{metadata !"", metadata !12}
!12 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 10, i32 -1, i32 0, i32 -1}
!13 = metadata !{metadata !"gl_PatchVerticesIn", i32 1, i32* @gl_PatchVerticesIn_typeProxy, metadata !14}
!14 = metadata !{i32 0, i32 3, i32 1088, null, i32 0, i32 34, i32 -1, i32 0, i32 -1}
!15 = metadata !{metadata !"gl_PrimitiveID", i32 1, i32* @gl_PrimitiveID_typeProxy, metadata !16}
!16 = metadata !{i32 0, i32 3, i32 1089, null, i32 0, i32 31, i32 -1, i32 0, i32 -1}
!17 = metadata !{metadata !"gl_TessCoord", i32 1, <3 x float>* @gl_TessCoord_typeProxy, metadata !18}
!18 = metadata !{i32 0, i32 3, i32 1090, null, i32 0, i32 37, i32 -1, i32 0, i32 -1}
!19 = metadata !{metadata !"gl_TessLevelOuter", i32 1, [4 x float]* @gl_TessLevelOuter_typeProxy, metadata !20}
!20 = metadata !{i32 0, i32 3, i32 1091, null, i32 3, i32 35, i32 -1, i32 0, i32 -1}
!21 = metadata !{metadata !"gl_TessLevelInner", i32 1, [2 x float]* @gl_TessLevelInner_typeProxy, metadata !22}
!22 = metadata !{i32 0, i32 3, i32 1095, null, i32 3, i32 36, i32 -1, i32 0, i32 -1}
!23 = metadata !{metadata !"patchIn", i32 1, <4 x float>* @patchIn_typeProxy, metadata !24}
!24 = metadata !{i32 0, i32 3, i32 1099, null, i32 3, i32 0, i32 -1, i32 0, i32 -1}
!25 = metadata !{metadata !"inb", i32 1, [32 x <2 x float>]* @inb_typeProxy, metadata !26}
!26 = metadata !{i32 0, i32 3, i32 1100, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!27 = metadata !{metadata !"ind", i32 1, [32 x <2 x float>]* @ind_typeProxy, metadata !28}
!28 = metadata !{i32 0, i32 3, i32 1132, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!29 = metadata !{metadata !"blb", i32 17, [32 x %testblb]* @blb_typeProxy, metadata !30, metadata !31}
!30 = metadata !{i32 0, i32 0, i32 1164, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!31 = metadata !{metadata !"testblb", metadata !8, metadata !"f", metadata !32}
!32 = metadata !{metadata !"", metadata !33}
!33 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!34 = metadata !{metadata !"bld", i32 17, [32 x %testbld]* @bld_typeProxy, metadata !35, metadata !36}
!35 = metadata !{i32 0, i32 0, i32 1196, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!36 = metadata !{metadata !"testbld", metadata !8, metadata !"f", metadata !32}
!37 = metadata !{metadata !"ivla", i32 1, [32 x <4 x float>]* @ivla_typeProxy, metadata !38}
!38 = metadata !{i32 0, i32 3, i32 23, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!39 = metadata !{metadata !"ivlb", i32 1, [32 x <4 x float>]* @ivlb_typeProxy, metadata !40}
!40 = metadata !{i32 0, i32 3, i32 24, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!41 = metadata !{metadata !"pinbi", i32 17, %pinbn* @pinbi_typeProxy, metadata !42, metadata !43}
!42 = metadata !{i32 0, i32 0, i32 1228, null, i32 3, i32 0, i32 -1, i32 0, i32 -1}
!43 = metadata !{metadata !"pinbn", metadata !8, metadata !"a", metadata !32}
!44 = metadata !{metadata !"centr", i32 1, [32 x <3 x float>]* @centr_typeProxy, metadata !45}
!45 = metadata !{i32 0, i32 3, i32 1230, null, i32 512, i32 0, i32 -1, i32 0, i32 -1}
!46 = metadata !{metadata !"", i32 16, %gl_PerVertex.0* @"anon@1_typeProxy", metadata !47, metadata !7}
!47 = metadata !{i32 0, i32 0, i32 1097, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!48 = metadata !{metadata !"ovla", i32 7, [2 x <4 x float>]* @ovla_typeProxy, metadata !38}
!49 = metadata !{metadata !"myColor2", i32 7, <3 x float>* @myColor2_typeProxy, metadata !50}
!50 = metadata !{i32 0, i32 3, i32 1229, null, i32 512, i32 0, i32 -1, i32 0, i32 -1}
#version 310 es
// LunarGOO output
#extension GL_ANDROID_extension_pack_es31a : enable
#extension GL_EXT_geometry_shader : enable
#extension GL_EXT_gpu_shader5 : enable
#extension GL_EXT_primitive_bounding_box : enable
#extension GL_EXT_shader_io_blocks : enable
#extension GL_EXT_tessellation_shader : enable
#extension GL_EXT_texture_buffer : enable
#extension GL_EXT_texture_cube_map_array : enable
#extension GL_KHR_blend_equation_advanced : enable
#extension GL_OES_sample_variables : enable
#extension GL_OES_shader_image_atomic : enable
#extension GL_OES_shader_multisample_interpolation : enable
#extension GL_OES_tessellation_point_size : enable
#extension GL_OES_texture_storage_multisample_2d_array : enable
layout(quads) in;
layout(fractional_odd_spacing) in;
layout(point_mode) in;
patch in highp vec4 patchIn;
in highp vec2 inb[32];
in highp vec2 ind[32];
in testblb {
	highp int f;
} blb[32];
in testbld {
	highp int f;
} bld[32];
layout(location=23) in highp vec4 ivla[32];
layout(location=24) in highp vec4 ivlb[32];
patch in pinbn {
	highp int a;
} pinbi;
in highp vec3 centr[32];
layout(location=23) out highp vec4 ovla[2];
out highp vec3 myColor2;

void main()
{
	gl_Position = gl_in[1].gl_Position;
	gl_PointSize = gl_in[1].gl_PointSize;
	
}

#version 310 es
// LunarGOO output
#extension GL_ANDROID_extension_pack_es31a : enable
#extension GL_EXT_geometry_shader : enable
#extension GL_EXT_gpu_shader5 : enable
#extension GL_EXT_primitive_bounding_box : enable
#extension GL_EXT_shader_io_blocks : enable
#extension GL_EXT_tessellation_shader : enable
#extension GL_EXT_texture_buffer : enable
#extension GL_EXT_texture_cube_map_array : enable
#extension GL_KHR_blend_equation_advanced : enable
#extension GL_OES_sample_variables : enable
#extension GL_OES_shader_image_atomic : enable
#extension GL_OES_shader_multisample_interpolation : enable
#extension GL_OES_tessellation_point_size : enable
#extension GL_OES_texture_storage_multisample_2d_array : enable
layout(quads) in;
layout(fractional_odd_spacing) in;
layout(point_mode) in;
patch in highp vec4 patchIn;
in highp vec2 inb[32];
in highp vec2 ind[32];
in testblb {
	highp int f;
} blb[32];
in testbld {
	highp int f;
} bld[32];
layout(location=23) in highp vec4 ivla[32];
layout(location=24) in highp vec4 ivlb[32];
patch in pinbn {
	highp int a;
} pinbi;
in highp vec3 centr[32];
layout(location=23) out highp vec4 ovla[2];
out highp vec3 myColor2;

void main()
{
	gl_Position = gl_in[1].gl_Position;
	gl_PointSize = gl_in[1].gl_PointSize;
	
}

